control.lua

```diff
@@ -156,165 +158,205 @@ end
 
 -- POWER MANAGEMENT --
 
--- Don't mess with this unless you mess with prototypes/entity/component.lua too.
--- Every number needs to correspond to a valid indicator entity name
-local VALID_POWER_TRANSFER_RATES = {1,2,5,10,20,50,100,200,500,1000,2000,5000,10000,20000,50000,100000} -- MW
-
-local function make_valid_transfer_rate(rate)
-	for _,v in pairs(VALID_POWER_TRANSFER_RATES) do
-		if v == rate then return v end
+function power_middleman_surface()
+	if game.surfaces["factory-power-connection"] then
+		return game.surfaces["factory-power-connection"]
+	end
+	
+	if #game.surfaces == 256 then
+		error("Unfortunately you have no available surfaces left for Factorissimo2. You cannot use Factorissimo2 on this map.")
 	end
-	return 0 -- Catchall
+	
+	local map_gen_settings = {height=1, width=1, property_expression_names={}}
+	map_gen_settings.autoplace_settings = {
+		["decorative"] = {treat_missing_as_default=false, settings={}},
+		["entity"] = {treat_missing_as_default=false, settings={}},
+		["tile"] = {treat_missing_as_default=false, settings={["out-of-map"]={}}},
+	}
+	
+	local surface = game.create_surface("factory-power-connection", map_gen_settings)
+	surface.set_chunk_generated_status({0, 0}, defines.chunk_generated_status.entities)
+	surface.set_chunk_generated_status({-1, 0}, defines.chunk_generated_status.entities)
+	surface.set_chunk_generated_status({0, -1}, defines.chunk_generated_status.entities)
+	surface.set_chunk_generated_status({-1, -1}, defines.chunk_generated_status.entities)
+	
+	return surface
 end
 
-local function update_power_settings(factory)
-	if factory.built then
-		local layout = factory.layout
-		-- Inside sender
-		local new_ies = factory.inside_surface.create_entity{
-			name = "factory-power-output-2-" .. factory.transfer_rate,
-			position = {factory.inside_x + layout.inside_energy_x, factory.inside_y + layout.inside_energy_y},
-			force = factory.force
-		}
-		new_ies.destructible = false
-		new_ies.operable = false
-		new_ies.rotatable = false
-		if factory.inside_energy_sender.valid then
-			factory.inside_energy_sender.destroy()
+local function remove_direct_connection(factory)
+	local dc = factory.direct_connection
+	if not dc or not dc.valid then return end
+	
+	for _, pole in pairs(factory.inside_power_poles) do
+		for _, neighbour in pairs(pole.neighbours.copper) do
+			if neighbour == dc then
+				local old = {}
+				for _, neighbour in ipairs(dc.neighbours.copper) do
+					if neighbour ~= pole then old[#old+1] = neighbour end
+				end
+				dc.disconnect_neighbour()
+				for _, neighbour in ipairs(old) do
```

Comment:
Using ipairs is creating a lot of weird code in this PR, such as the old[#old + 1],

It would be cleaner to use pairs, and to index by the poles unit number